File: problems/2000_Reverse_Prefix_of_Word/solution.go
package main

import (
	"fmt"
	"strings"
);

func reversePrefix(word string, ch byte) string {
	reversed := ""

	firstOccurence := strings.IndexByte(word, ch)

	if firstOccurence < 0 {
		return word
	}

	for i := firstOccurence; i >= 0; i-- {
		reversed += string(word[i])
	}

	reversed += word[firstOccurence + 1:]

	return reversed;
};

/**
 * Time complexity: O(n + m)
 * Space complexity: O(1)
 */

func main() {
	fmt.Println(reversePrefix("abcdefd", 'b'));
	fmt.Println(reversePrefix("xyxzxe", 'z'));
	fmt.Println(reversePrefix("abcd", 'z'));
}

File: problems/2000_Reverse_Prefix_of_Word/solution.ts
export {}

function reversePrefix(word: string, ch: string): string {
  let reversed = ""

  const firstOccurence = word.indexOf(ch)

  if (firstOccurence < 0) {
    return word
  }

  for (let i = firstOccurence; i >= 0; i--) {
    reversed += word[i]
  }

  reversed += word.substring(firstOccurence + 1, word.length)

  return reversed
}

/**
 * Time complexity: O(n + m)
 * Space complexity: O(1)
 */

console.log(reversePrefix("abcdefd", "d"))
console.log(reversePrefix("xyxzxe", "z"))
console.log(reversePrefix("abcd", "z"))

function reversePrefix2(word: string, ch: string): string {
  let res = word

  for (let i = 0; i < word.length; i++) {
    if (word[i] !== ch) continue

    const reversed = word
      .substring(0, i + 1)
      .split("")
      .reverse()
      .join("")

    res = `${reversed}${word.substring(i + 1, word.length)}`

    break
  }

  return res
}

/**
 * Time complexity: O(n + 3n) (?)
 * Space complexity: O(1)
 */

console.log(reversePrefix2("abcdefd", "d"))
console.log(reversePrefix2("xyxzxe", "z"))
console.log(reversePrefix2("abcd", "z"))
